### ISSUE TITLE
RangeError: Incorrect locale information provided - causing Vercel Edge middleware function to fail

### ISSUE THREAD

**Issue by Sam-Scolari** [i18nexus/next-i18n-router#25]
Next.js Version: 13.5.4 (latest)
next-i18n-router Version: 4.1.1

I have been having [SEO issues](https://stackoverflow.com/questions/77247550/open-graph-not-picking-up-on-nextjs-app-seo-meta-tags) recently on my NextJS app and I am inclined to believe that this edge function error in my middleware might be the culprit. The i18n routing in my app works fine except for this 500 error for the / route that seems to get thrown when I first deploy my app. Any ideas on how I could solve this and if this error could be the reason I am having these issues?

![image](https://github.com/i18nexus/next-i18n-router/assets/56174573/6538c9e1-6d7c-4f43-8044-d5af4fa6f9f7)
![image](https://github.com/i18nexus/next-i18n-router/assets/56174573/886dccb4-873a-48c7-8a61-d04e774c73ec)

middleware.js
```js
import { userAgent } from "next/server";
import { i18nRouter } from "next-i18n-router";
import i18nConfig from "./i18nConfig";

export function middleware(request) {
  let i18nRequest = i18nRouter(request, i18nConfig);

  // Get the device type
  const device = getDevice(request);
  i18nRequest.headers.set("x-device-type", device);

  return i18nRequest;
}

// only apply this middleware to files in the app directory
export const config = {
  matcher: "/((?!api|static|.*\\..*|_next).*)",
};

function getDevice(request) {
  const { device } = userAgent(request);

  return device.type === "mobile" ? "mobile" : "desktop";
}
```

i18nConfig.js
```js
const i18nConfig = {
  locales: ["en", "pt"],
  defaultLocale: "en",
  routingStrategy: "dynamicSegment",
};

module.exports = i18nConfig;
```

Here is the full error message
```bash
RangeError: Incorrect locale information provided
    at (node_modules/next-i18n-router/node_modules/@formatjs/intl-localematcher/lib/abstract/CanonicalizeLocaleList.js:7:16)
    at (node_modules/next-i18n-router/node_modules/@formatjs/intl-localematcher/lib/index.js:8:34)
    at (node_modules/next-i18n-router/dist/localeDetector.js:12:43)
    at (node_modules/next-i18n-router/dist/i18nRouter.js:37:25)
    at (middleware.js:6:31)
    at (node_modules/next/dist/esm/server/web/adapter.js:163:22)
    at (node_modules/next/dist/esm/server/async-storage/request-async-storage-wrapper.js:72:23)
    at (node_modules/next/dist/esm/server/web/adapter.js:150:52)
```

**Comment by i18nexus** [comment]
**tldr; We have made a quick patch with some tests to fix this and it is now released in version 4.1.2.** 

After some quick research I discovered this post: 
https://stackoverflow.com/questions/76447732/nextjs-13-i18n-incorrect-locale-information-provided

It appears the the `match` function from `@formatjsintl-localematcher` fails when no accept-language header is present on the request. This function is used in `i18nRouter`'s default locale detection. 

If bots don't have an accept-language header on their requests, it's likely this is the cause of the SEO issue you're experiencing.

`@formatjsintl-localematcher` is the recommended library to use for locale detection in the Next.js i18n docs. I should probably make a PR to Next.js docs to add a note about this behavior.

Thanks for the report!

**Comment by Sam-Scolari** [comment]
You were right, updating solved my issue! Thanks for the quick reply!

**Comment by faheempatel** [comment]
I am seeing this error in 5.0.1 :(
Any help would be greatly appreciated!

**Comment by i18nexus** [comment]
Hi @faheempatel. Can you expand on what you're experiencing?

**Comment by faheempatel** [comment]
> Hi @faheempatel. Can you expand on what you're experiencing?

Hi there, I am seeing this error being reported in Sentry under `middleware.ts` and the app itself is deployed on Vercel.

**Comment by faheempatel** [comment]
@i18nexus Just bumping this up in case the notification was missed. Appreciate the help 🙏🏼

**Comment by i18nexus** [comment]
Can you show me your i18n config?

**Comment by faheempatel** [comment]
> Can you show me your i18n config?

```javascript
export const i18nConfig = {
  locales: ['en', 'ar'],
  defaultLocale: 'en',
};
```

**Comment by i18nexus** [comment]
Can you share the error message?On Nov 15, 2023, at 3:25 AM, Faheem Patel ***@***.***> wrote:﻿

Can you show me your i18n config?

export const i18nConfig = {
  locales: ['en', 'ar'],
  defaultLocale: 'en',
};

**Comment by faheempatel** [comment]
There isn't much of a stack trace for me to look at at the moment (waiting for the error to occur again) but this is what I saw on Sentry under the middleware:
![2023-11-17 at 11 01 06](https://github.com/i18nexus/next-i18n-router/assets/458311/695c0d74-280a-460d-8b2c-62bb57f661d8)

**Comment by i18nexus** [comment]
@faheempatel This is now fixed in version v5.0.2

After further research, it appears that `@formatjsintl-localematcher` is picky about the type of locales it will use from the accept-language header. For example, providing a locale of "asdf" will cause the RangeError you're experiencing.

It is strange that it throws an error instead of falling back to the defaultLocale. Obviously we do not want our app to crash if someone visits our site with an accept-language header that the localeMatcher finds to be invalid. To fix this, we have made an update to our default `localeDetector` in version 5.0.2 so that when the matcher throws an error, the error will be caught and returns the defaultLocale. This will put an end to this error once and for all. In the meantime we're looking into alternatives to `@formatjsintl-localematcher` since catching errors isn't a very elegant solution. But everything should work as expected now 👍

**Comment by faheempatel** [comment]
@i18nexus Thank you! Really appreciate all your help and time!
